Add MathML support

adamsilverstein opened the pull request:
Implements MathML support via an Iframe and the CDN hosted MathJax script. Fixes https://github.com/ampproject/amphtml/issues/12800

* resizes the iframe height after the math formula renders
* ideally the iframes should also scale their width so smaller formulas could be embedded inline.

Includes a few example of formulas, I wasn't sure how to make these small to begin with, in my testing they started large then resized smaller when the formula rendered.

![](https://cl.ly/043I2d380o0M/download/Image%202018-01-12%20at%208.20.20%20PM.png)

adamsilverstein commented:
Hey @googlebot - I signed it!

honeybadgerdontcare commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,69 @@
+#
```
Please take a look at [Contributing Validator Rules](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md). The format for allowing the script tag has been simplified greatly.

adamsilverstein replied:
Ok, will do - thanks for the feedback.

adamsilverstein replied:
I fixed the extensions/amp-mathml/validator-amp-mathml.protoascii file in https://github.com/ampproject/amphtml/pull/12836/commits/f6d93f3a5b14cdc926274eeafdc36405f11d4c6d - I still need to add some valid and invalid test cases

aghassemi commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+      document.body.appendChild(div);
+      MathJax.Hub.Queue( function () {
+        const rendered = document.getElementById('MathJax-Element-1-Frame');
+        context.requestResize(
```
`updateDimensions` is okay in this case ( `requestResize` is not guaranteed to resize). We make exceptions for `updateDimensions` per component and this should be fine.

aghassemi replied:
also `mathml` needs to be added to `updateDimensionsEnabled_()` method in `ampcontext-integration.js` file for it to work.

adamsilverstein replied:
interesting, will add

adamsilverstein replied:
@aghassemi when adding this I noticed on a line I was touching that the check for 'github' isn't using a strict equality check, see https://github.com/ampproject/amphtml/blob/master/3p/ampcontext-integration.js#L72

This should be `===`, right? Should I open a separate pull request for this issue? I don't want to introduce unrelated changes here.

aghassemi replied:
Our JS compiler handles discrepencies between `==` and `===` so it is fine in practices, would be nice for consistency, you can change it here in this PR

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.js:
```diff
@@ -0,0 +1,82 @@
+  /** @override */
+  buildCallback() {
+    const formula = this.element.getAttribute('formula');
+    if (!formula || '' === formula) {
```
let's have 
```
  user().assert(
       data.formula,
       'The formula attribute is required for <amp-mathml> %s',
       data.element);
```
here instead.

adamsilverstein replied:
ok!

adamsilverstein replied:
that bit is in mathml.js, i removed the buildCallback entirely, it seems superfluous

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.js:
```diff
@@ -0,0 +1,82 @@
+    super(element);
+
+    /** @private {!Element} */
+    this.container_ = this.win.document.createElement('div');
```
not used, remove

adamsilverstein replied:
ok

aghassemi commented on validator/testdata/feature_tests/amp-mathml.html:
```diff
@@ -0,0 +1,27 @@
+<!doctype html>
```
keep this in the extension folder. Also needs a `.out` file. See https://github.com/ampproject/amphtml/blob/master/extensions/amp-sidebar/0.1/test/validator-amp-sidebar.html and https://github.com/ampproject/amphtml/blob/master/extensions/amp-sidebar/0.1/test/validator-amp-sidebar.out as a good sample.

adamsilverstein replied:
ok, i have an uncommitted out file that i think was generated when i started, I'll check that and commit it.

adamsilverstein replied:
I added the file, but I am still missing something.

honeybadgerdontcare replied:
See here for an example [extensions](https://github.com/ampproject/amphtml/tree/master/extensions/amp-gfycat/0.1/test) test directory with an .html and .out.

aghassemi commented on yarn.lock:
```diff
@@ -178,6 +178,12 @@ acorn-globals@4.1.0, acorn-globals@^4.0.0:
   dependencies:
     acorn "^5.0.0"
 
+acorn-globals@^3.1.0:
```
revert this file

adamsilverstein replied:
ok, will do

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+      MathJax.Hub.Queue( function () {
+        const rendered = document.getElementById('MathJax-Element-1-Frame');
+        context.requestResize(
+                rendered./*OK*/offsetWidth + 20,
```
i think i was getting cut off display, i'll retest and remove or document reason

adamsilverstein replied:
My display was cut off when removing the extra height (https://cl.ly/1G243K2x3W3X) - I tracked this down to same margins built into the formulas rendered by MathJax. I added some code removing them on the fly in https://github.com/ampproject/amphtml/pull/12836/commits/70e7b22601609c53227c095d2b159c81d56f16d5. Do I need to add more resiliency here, checking that I found the element before applying styles?

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
oh yea; sorry, i started by copying an existing component, I'll clean this up.

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+console.log( 'mathml.js' );
```
:doh:

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+ */
+function getMathmlJs(global, scriptSource, cb) {
+ writeScript(global, scriptSource, function() {
+    cb(global.gist);
```
right!

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+import {user} from '../src/log';
+
+/**
+ * Get the correct script for the gist.
```
yes, thanks

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+      delete data.height;
+      const div = document.createElement('div');
+      div.setAttribute('id','mathmlformula');
+      div.innerHTML = data.formula;
```
interesting, ok i'll change and do some testing

adamsilverstein replied:
So far my test cases look fine after the change to `textContent`, I'll add some more samples from https://developer.mozilla.org/en-US/docs/Mozilla/MathML_Project/MathML_Torture_Test to my example page

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+      const div = document.createElement('div');
+      div.setAttribute('id','mathmlformula');
+      div.innerHTML = data.formula;
+      document.body.appendChild(div);
```
ok, will change - what is the context/reason behind `global.`?

aghassemi replied:
It may not be the current window in certain situations where an AMP doc is embedded differently.

adamsilverstein commented:
@aghassemi thanks for the careful review and detailed feedback. I will work on all of the points raised and get back to you when I am ready for another review.

I still have a question about the resizing: is it possible to get the iframe to resize its width as well as its height, currently my listener only changes height. any tips to getting this working?

adamsilverstein commented on 3p/mathml.js:
```diff
@@ -0,0 +1,68 @@
+  getMathmlJs(
+    global,
+    'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML',
+    function() {
```
got it

aghassemi commented:
@adamsilverstein There are a few options regarding resizability of this component that I am running by the team in a [review meeting](https://github.com/ampproject/amphtml/issues/12551#issuecomment-358375447) today, I will get back to you on the decision.

adamsilverstein commented:
@aghassemi excellent, thanks for your support on this PR! 

Hopefully, I addressed all of the issues you raised in my recent commits except the tests piece, I think I just need to move the file and work on adding some tests? (https://github.com/ampproject/amphtml/pull/12836/files#r161914420).

adamsilverstein commented:
> Reading resizing, we only resize with and AMP components are block elements by default. User would need to use flex or other parent CSS ways to make this inline. We won't do resizing for width.

@aghassemi I missed this comment earlier, can you elaborate on how this would be achieved or point me to an example? Remember, we don't know the width of the formula until it renders :)

aghassemi commented:
@adamsilverstein The team had a good chat regarding this component today and the conclusion is this:

1- We should restrict the `layout` attribute of this component to `container` only. Idea is that this is an intrinsically sized component and it will always resize itself to wait as much or little space as needed. Author should not worry about defining any `width/height` for it.
2- We should add a new attribute called `inline` which would controls whether the component is `display: inline-block` or `container.

I will add some comments in the code there changes are needed for this. 

Any thoughts?

aghassemi commented on 3p/ampcontext-integration.js:
```diff
@@ -69,7 +69,8 @@ export class IntegrationAmpContext extends AbstractAmpContext {
     return (this.embedType_ === 'facebook'
         || this.embedType_ === 'twitter'
-        || this.embedType_ == 'github');
+        || this.embedType_ == 'github'
```
as you mentioned, let's change the `==` to be consistent.

adamsilverstein replied:
right, fixed in https://github.com/ampproject/amphtml/pull/12836/commits/d1bfee5726bed9c061e5e1c1151bc247d8488464

aghassemi commented on examples/amp-mathml.amp.html:
```diff
@@ -0,0 +1,29 @@
+</head>
+<body>
+  <h2>The Quadratic Formula</h2>
+  <amp-mathml
```
all samples now should now only do `layout=container`

aghassemi replied:
let's also add one sample that's inline in a sentence.

adamsilverstein replied:
added

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.js:
```diff
@@ -0,0 +1,72 @@
+
+  /** @override */
+  isLayoutSupported (layout) {
+    return isLayoutSizeDefined(layout);
```
`return layout == Layout.CONTAINER;`

adamsilverstein replied:
right, changed

aghassemi commented on gulpfile.js:
```diff
@@ -112,6 +112,7 @@ declareExtension('amp-lightbox', '0.1', true);
 declareExtension('amp-lightbox-viewer', '0.1', true);
 declareExtension('amp-list', '0.1', false);
 declareExtension('amp-live-list', '0.1', true);
+declareExtension('amp-mathml', '0.1', false);
```
Now we need a `CSS` file as well and `false` becomes `true`
The CSS file would handle the `inline` attribute by simply doing 
```
amp-mathml[inline] {
  display: inline-block;
}
```

Note: we _may_ need to do 
```
amp-mathml {
  height: 1px;
}
amp-mathml[inline] {
  width: 1px;
}
```
to make it not be 0by0 until resize happens. Try it without the above 1px hacks, if it works all is good. Otherwise if you see mathML not rendering, we may need to add them. (This is because AMP tends to ignore elements that are 0x0)

adamsilverstein replied:
The 1px settings were required, without that the iframe never rendered.

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.js:
```diff
@@ -0,0 +1,72 @@
+    this.applyFillContent(iframe);
+    // Triggered by context.updateDimensions() inside the iframe.
+    listenFor(iframe, 'embed-size', data => {
+      this./*OK*/changeHeight(data['height']);
```
this can become 
`this.element.getResources()./*OK*/changeSize(
        this.element, newHeight, /* newWidth */ newWidth);`

But I believe we should not change `width` unless it is `inline`,  so give the following a try first:

```
if(!this.element.hasAttribute('inline') ) {
 // Don't change the width if not inlined.
 newWidth = undefined;
} 
this.element.getResources()./*OK*/changeSize(
        this.element, newHeight, /* newWidth */ newWidth);
```

adamsilverstein replied:
ok, got it. currently my iframes aren't rendering locally so hard to test; this change looks good to me.

adamsilverstein replied:
This works great, thanks!

adamsilverstein commented:
@aghassemi thanks for all the feedback and especially the effort and assistance getting the resizing to work. This looks really good in my testing, the inline formula rendering works great:

![](https://cl.ly/3d0e1e2f1f0L)

This is exactly the type of improved display I was hoping to get out of this PR vs. our existing implementation (which uses amp-iframes). The ability to display math formulas inline is a big improvement - especially for shorter formulas.

I think the only thing missing at this point is improvements for the test/out pages?

aghassemi commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,44 @@
+  html_format: AMP
+  tag_name: "AMP-mathml"
+  requires_extension: "amp-mathml"
+  attrs: {
```
remove

adamsilverstein replied:
removed

aghassemi commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,44 @@
+    name: "data-selected-cat"
+    value_regex_casei: "(oscar|chloe|bella)"
+  }
+  attrs: {
```
add `inline`

adamsilverstein replied:
added

aghassemi commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,44 @@
+  amp_layout: {
+    supported_layouts: CONTAINER
+  }
+}
```
new line at the end of file

adamsilverstein replied:
added. out of curiosity, why is this required? also, why doesn't the linter pick up this as an issue?

aghassemi commented on validator/testdata/feature_tests/amp-mathml.html:
```diff
@@ -0,0 +1,27 @@
+    <h2>Cauchy's Integral Formula</h2>
+    <amp-mathml layout="responsive" formula="\[f(a) = \frac{1}{2\pi i} \oint\frac{f(z)}{z-a}dz\]">
+    </amp-mathml>
+    <h2>Double angle formula for Cosines</h2>
```
add one with `inline`

adamsilverstein replied:
added, copied over all of items from examples.

aghassemi commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,44 @@
+  attr_lists: "common-extension-attrs"
+}
+
+tags: {  # <amp-mathml>
```
/to @honeybadgerdontcare

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.css:
```diff
@@ -0,0 +1,27 @@
+
+amp-mathml[inline] {
+  width: 1px;
+}
```
new line at the end of the file.

adamsilverstein replied:
added

aghassemi commented on extensions/amp-mathml/0.1/amp-mathml.js:
```diff
@@ -0,0 +1,76 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {isLayoutSizeDefined} from '../../../src/layout';
```
remove

adamsilverstein replied:
removed

aghassemi commented on extensions/amp-mathml/0.1/test/test-amp-mathml.js:
```diff
@@ -0,0 +1,32 @@
+ * limitations under the License.
+ */
+
+describes.realWin('amp-mathml', {
```
for testing, I recommend writing end-2-end integration tests. https://github.com/ampproject/amphtml/blob/master/extensions/amp-sidebar/0.1/test/integration/test-amp-sidebar.js is a good one to use as reference.

In short, the test will essentially render an AMP page with bunch of `amp-mathml` tags you define. Then inside of the test, use various utilities like `poll` to wait for those `math-ml` tags to "render" and "resize". Unfortunately because they are rendered in an iframe, it is not possible to assert that MathML itself rendered but at minimum you can assert that iframe was created and resized to something bigger than 1px.

aghassemi commented:
@adamsilverstein Looks awesome! Thanks for all your work on this. Left some comments regarding a testing strategy. Most of remaining work is just polish:
- [ ] tests
- [ ] readme file clean up
- [ ] handling final rounds of review nit pick comments
- [ ] lint and getting tests to be green
- [ ] merge :tada:

adamsilverstein commented:
Thanks again for the detailed feedback @aghassemi - I'll work on the requested changes and get back to you if I have any questions.

adamsilverstein commented:
>  readme file clean up

Took a pass at the readme, can you please review: https://github.com/adamsilverstein/amphtml/blob/0e3dc5f4d21f5037caea06cdfc8c5e25941c558f/extensions/amp-mathml/amp-mathml.md

adamsilverstein commented:
not sure how to clear this validator error, tips appreciated! where do i add amp-mathml as an allowed tag?

`feature_tests/amp-mathml.html:14:2 Only AMP runtime 'script' tags are allowed, and only in the document head. (see https://www.ampproject.org/docs/reference/spec#html-tags) [CUSTOM_JAVASCRIPT_DISALLOWED]`

adamsilverstein commented:
@aghassemi -
I took a first pass at writing some tests. This is the first time I have tried to write this type of integration test and I would love some help getting it right. Feel free to jump in and take over the PR to get the tests finished and validation working correctly (see my comment above).

Regarding the integration tests: we need to wait for both the iframe to render, and ideally for the MathJax script to fire and complete rendering the formula. Since we already wait for the MathJax rendering inside the iframe and then trigger `context.requestResize` with the correct dimensions, I'm wondering if we should listen for this event firing in the poll or as the test itself? 

Thanks!

aghassemi commented:
@adamsilverstein I can take over writing tests in later PRs. If you don't mind, let's fix the lint issues and the last validator comment and merge the PR. Also if you like, please add a OWNER.yaml file to the folder with your github account. 

Thanks again for all your impressive work on this component.

adamsilverstein commented on extensions/amp-mathml/validator-amp-mathml.protoascii:
```diff
@@ -0,0 +1,43 @@
+
+tags: {  # <amp-mathml>
+  html_format: AMP
+  tag_name: "AMP-mathml"
```
Thanks for reviewing @honeybadgerdontcare - correctred in https://github.com/ampproject/amphtml/pull/12836/commits/6fdb6e4d6fa7697e347954537ff0617f8f469e68

adamsilverstein commented:
@aghassemi You are very welcome, thanks for your assistance along the way. Adding tests later sounds good, please tag me in the PR as I would like to see and learn from how you add these.

I fixed the linting errors and added the OWNERS.yaml file as well as changing the tagname to allcaps as suggested by @honeybadgerdontcare.

Did you see my question above about clearing the valtador errors (https://github.com/ampproject/amphtml/pull/12836#issuecomment-359198630) - unless the allcaps change in https://github.com/ampproject/amphtml/pull/12836/commits/6fdb6e4d6fa7697e347954537ff0617f8f469e68 fixes them, I'm not sure how to correct those. On a related note, I'm blocked from running `gulp validator` locally - when I do, I hit an error I haven't been able to resolve: `google.protobuf.descriptor not found. Try "apt-get install python-protobuf"` - any suggestions for getting around that, I'm on the latest mac osx.

adamsilverstein commented:
@aghassemi I think I have fixed all the warnings other than the validator messages I mentioned above that I am not sure how to clear. Please let me know how to fix or just go for it and get the validator passing.

Thanks!

adamsilverstein commented:
missed the earlier comment from @honeybadgerdontcare - move the test and out files into the test folder in https://github.com/ampproject/amphtml/pull/12836/commits/98d003979f57758aaca472254544aa52dfa35cf4

honeybadgerdontcare commented:
@adamsilverstein it looks like you're still missing the `.out` file. If this is inconvenient please open an issue assigned to me with your current `.html` test file and I'll take care of the rest. That will get you around any travis/test issues.

adamsilverstein commented:
@honeybadgerdontcare I created an issue but don't have the capability to assign it to you: https://github.com/ampproject/amphtml/issues/12973

honeybadgerdontcare commented:
I've taken the issue. The problem, or at least the easiest to see, is with the out file. It doesn't include the actual `.html` contents. See [this](https://github.com/ampproject/amphtml/blob/master/extensions/amp-facebook-like/0.1/test/validator-amp-facebook-like.out) as an example. Regardless, I can take care of it for you in a separate PR. So remove the test html and out files from this PR.

adamsilverstein commented:
@honeybadgerdontcare thanks so much for your assistance, I removed the test files in https://github.com/ampproject/amphtml/pull/12836/commits/85bf03906ffe7be6362638824a0f029bd2742e87.

adamsilverstein commented:
@aghassemi all checks have passed :)

aghassemi commented:
@adamsilverstein merged! I am super happy to see large contributions like this from the community. Thanks again Adam.

/cc @mrjoro

adamsilverstein commented:
Wooohooo! Thanks for all your shepherding @aghassemi I learned a ton about AMP along the way, and this feature will be a big improvement for our client's site. 

How long does it take for this feature to make it to production/the wild?

aghassemi commented:
@adamsilverstein For production use, ~3 weeks (Validator changes take a bit to make it to prod. A comment will be added to this PR when validation rules are in prod). Folks can start experimenting with it in next Dev Channel (tomorrow) however.
